### PR TITLE
ENH: new MultiPlotter class

### PR DESCRIPTION
**Pull request by GuillaumeFavelier**
Mainly proof of concept `Plotter` using separated instances of `QtInteractor` to solve isolation issues with SSAO and EDL too.

image

Related to https://github.com/mne-tools/mne-python/pull/8771#issuecomment-765563603

<details>
<summary>Code snippet</summary>

```py
import numpy as np
import vtk
import pyvistaqt as pvqt
from pyvista import examples


def _enable_ssao(plotter, dataset=None, idx=None):
    lightsP = vtk.vtkLightsPass()
    opaqueP = vtk.vtkOpaquePass()
    translucentP = vtk.vtkTranslucentPass()
    volumeP = vtk.vtkVolumetricPass()

    collection = vtk.vtkRenderPassCollection()
    collection.AddItem(lightsP)

    # opaque passes
    ssaoCamP = vtk.vtkCameraPass()
    ssaoCamP.SetDelegatePass(opaqueP)

    if dataset is None:
        ssaoP = vtk.vtkSSAOPass()
        ssaoP.SetRadius(30)
        ssaoP.SetBias(0.1)
        ssaoP.SetKernelSize(32)
        ssaoP.SetBlur(True)
    else:
        bounds = np.asarray(dataset.GetBounds())
        scene_size = np.linalg.norm(bounds[:3] - bounds[3:])
        ssaoP = vtk.vtkSSAOPass()
        ssaoP.SetRadius(0.1 * scene_size)  # comparison radius
        ssaoP.SetBias(0.001 * scene_size)  # comparison bias
        ssaoP.SetKernelSize(128)  # number of samples used
        ssaoP.BlurOff()  # do not blur occlusion
    ssaoP.SetDelegatePass(ssaoCamP)
    collection.AddItem(ssaoP)

    # translucent and volumic passes
    ddpP = vtk.vtkDualDepthPeelingPass()
    ddpP.SetTranslucentPass(translucentP)
    ddpP.SetVolumetricPass(volumeP)
    collection.AddItem(ddpP)

    # finally overlays
    overP = vtk.vtkOverlayPass()
    collection.AddItem(overP)

    sequence = vtk.vtkSequencePass()
    sequence.SetPasses(collection)

    camP = vtk.vtkCameraPass()
    camP.SetDelegatePass(sequence)

    if idx is None:
        plotter.renderer.SetPass(camP)
    else:
        plotter.renderers[idx].SetPass(camP)


nefertiti = examples.download_nefertiti()
point_cloud = examples.download_lidar()
teapot = examples.download_teapot()
bunny = examples.download_bunny_coarse().triangulate()

p = pvqt.Plotter(shape=(2, 2))
p.interactors[0].add_mesh(nefertiti, color=True)
p.interactors[0].enable_eye_dome_lighting()
p.interactors[0].add_text("Eye-Dome Lighting", font_size=24)

p.interactors[1].add_mesh(bunny, color=True)
_enable_ssao(p.interactors[1], dataset=bunny)
p.interactors[1].add_text("SSAO", font_size=24)

p.interactors[2].add_mesh(point_cloud, color=True)
p.interactors[2].enable_eye_dome_lighting()
p.interactors[2].add_text("Eye-Dome Lighting", font_size=24)

p.interactors[3].add_mesh(teapot, color=True)
_enable_ssao(p.interactors[3], dataset=teapot)
p.interactors[3].add_text("SSAO", font_size=24)

p.show()
```

</details>

Closes https://github.com/pyvista/pyvistaqt/issues/7

### CODE REVIEW

**Comment by GuillaumeFavelier** [comment]
Thanks to 719219e it is now possible to embed multiple full-fledged `BackgroundPlotter`s in one `MultiPlotter`:

```py
import pyvistaqt as pvqt
from pyvista import examples

nefertiti = examples.download_nefertiti()
point_cloud = examples.download_lidar()
cow = examples.download_cow()
teapot = examples.download_teapot()

mp = pvqt.MultiPlotter(shape=(2, 2), window_size=(800, 600))

plotter = mp.select((0, 0))
plotter.add_mesh(nefertiti, color=True)

plotter = mp.select((0, 1))
plotter.add_mesh(point_cloud, color=True)

plotter = mp.select((1, 0))
plotter.add_mesh(cow)

plotter = mp.select((1, 1))
plotter.add_mesh(teapot)

mp.show()
```

image

**Comment by GuillaumeFavelier** [comment]
I added a new class called `MultiPlotter`. The API is quite simple, **it is not a plotter though**, it's just an interface (Personally, I see it as a multiplexer). One can get access to the internal plotters through `select()`. After that, the usage does differ. Parameters can be sent at init with `**kwargs`.

Finally, I added testing and did a bit of refactoring to avoid code redundancy between the plotters.

Looking at #7, the widgets are not draggable but I think the main issue is addressed. [I was able to confirm](https://github.com/mne-tools/mne-python/pull/8771#issuecomment-765563603) that isolating the plotter in this way mitigates the issue with EDL and SSAO (not yet implemented).

What do you think @pyvista/developers ?

**Comment by akaszynski** [comment]
Played around with this for a bit and I liked the interface.  It's helpful to have individual toolbars for each plotter.

One note.  The "q" key still works and pressing it deletes a frame/sub-plotter.  Is this intended?  Also, is it possible to dynamically add additional frames to the `MultiPlotter`?  If so, would we need a main window toolbar?

**Comment by GuillaumeFavelier** [comment]
Thanks for trying!

> One note. The "q" key still works and pressing it deletes a frame/sub-plotter. Is this intended?

Oh not at all :thinking: . I can mitigate by disabling the shortcuts in individual plotters for now. Maybe for long-term, it would best to have the shorcut works only on the plotter focused by the mouse.

>  Also, is it possible to dynamically add additional frames to the MultiPlotter? If so, would we need a main window toolbar?

What do you mean? It was not designed to support adding/removing plotters dynamically, mostly because of the `shape` parameter. The `MultiPlotter`'s itself can have a toolbar of course. I can add it if that's something you think could be interesting to have.

**Comment by akaszynski** [comment]
> What do you mean? It was not designed to support adding/removing plotters dynamically, mostly because of the `shape` parameter. The `MultiPlotter`'s itself can have a toolbar of course. I can add it if that's something you think could be interesting to have.

I only thought of this because by pressing "q" I could dynamically remove plotters and the window automatically resized.  Should you wish, I figured you could do this in the opposite direction.  For now, I say keep everything fixed and don't have a "global" toolbar on the main window.

Disabling the "q" shortcut seems to be the only thing left to do for this PR.  The rest is great!

**Comment by GuillaumeFavelier** [comment]
I'll work on that next and ping when I'm done 👍

**Comment by GuillaumeFavelier** [comment]
Now `allow_quit_keypress` is set to `False` by default. You may be interested @larsoner 

I'll work on the doc now

**Comment by GuillaumeFavelier** [comment]
This is ready to go on my end @pyvista/developers

File: docs/usage.rst [hunk 1]
```diff
@@ -42,6 +42,32 @@ and functions like a normal PyVista ``Plotter`` instance. For example:
...
+
+.. code:: python
+
+    import pyvista as pv
+    from pyvistaqt import MultiPlotter
+
+    mp = MultiPlotter(shape=(2, 2), window_size=(800, 600))
+
+    plotter = mp.select((0, 0))
+    plotter.add_mesh(pv.Sphere())
```
Comment (agramfort) [inline docs/usage.rst:59]:
```suggestion
    mp[0, 0].add_mesh(pv.Sphere())
```

?

Comment (GuillaumeFavelier) [reply docs/usage.rst:59]: Oh this is interesting!

**Comment by GuillaumeFavelier** [comment]
Using `__getitem__` simplifies usage a great deal!

```py
import pyvista as pv
from pyvistaqt import MultiPlotter

mp = MultiPlotter(shape=(2, 2))
mp[0].add_mesh(pv.Sphere())
mp[1].add_mesh(pv.Cylinder())
mp[2].add_mesh(pv.Cube())
mp[3].add_mesh(pv.Cone())
```

image

File: docs/usage.rst (hunk 1 above)
Comment (GuillaumeFavelier) [reply docs/usage.rst:59]: I'm sold to use `__getitem__` instead of `select()`. This comes for free and really helps usage.

**Comment by agramfort** [comment]
you should use mp[0, 1] ie 2 indices otherwise it's not clear which plotter you refer to.
See how matplotlib does it (squeeze parameter). So if you have nrows=1 or ncols=1 you can
use only one index

**Comment by GuillaumeFavelier** [comment]
> you should use mp[0, 1] ie 2 indices otherwise it's not clear which plotter you refer to.

I let the possibility for linear indexing but I can remove it if it's confusing.

Also `shape` is not captured by  `MultiPlotter` meaning that multi-plottter **and** multi-renderer is possible:

image

**Comment by GuillaumeFavelier** [comment]
The state of this PR has changed a bit since your last review @akaszynski, sorry for that:

* `allow_quit_keypress=False` by default
* `docs/usage` and `docs/api_ref` are updated with `MainPlotter`
* `MainPlotter`s `shape` became `nrows` and `ncols` with `__getitem__`/`__setitem__`

**Review by akaszynski — APPROVED** [review]
Looks good!  Thanks for adding this @GuillaumeFavelier.

**Comment by larsoner** [comment]
Thanks @GuillaumeFavelier !

**Pull request merged**